### DISCUSSION TITLE
kornia.warp_perspective() GPU support?

### DISCUSSION THREAD

**Discussion by pentadotddot** [kornia/kornia#1008] [Questions] [answered]
Hi!

I am struggling with real-time video stitching, and I found this kornia repo. It claims strong gpu support, altough I could not find any snippet to utilize gpu to do the warping. It would be really nice because I think warping is too intense for the cpu as it goes through all pixels, it would be great if it could utilize gpus. Any idea? Maybe it is already possible but I could not find how to do it...
It is really slow this way, slower then the opencv version...

**Comment by shijianjian** [comment, marked as answer]
It will be computed inside GPU if the image is a cuda tensor.

**Reply by pentadotddot** [reply]
Thank you for the quick reply.
So I tried it this way:

```python
device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
print(device)
def warp_image(img,M):
	img = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)
	# convert to torch tensor
	data: torch.tensor = kornia.image_to_tensor(img, keepdim=False)  # BxCxHxW

	data = torch.tensor(data, device=device).float()
	h,w = img.shape[1],img.shape[0]
	# warp the original image by the found transform
	data_warp: torch.tensor = kornia.warp_perspective(data, M, dsize=(h, w))

	# convert back to numpy
	img_warp: np.ndarray = kornia.tensor_to_image(data_warp.byte())

	img_warp: np.ndarray = kornia.tensor_to_image(data_warp.byte())

	return img_warp
	
```
But when I execute this defined function I get this error: 
stitch_benchmark.py:142: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).
  data = torch.tensor(data, device=device).float()
/home/mobilrouter/.conda/envs/Mobil_Torch/lib/python3.8/site-packages/kornia/geometry/transform/imgwarp.py:99: UserWarning: The align_corners default value has been changed. By default now is set True in order to match cv2.warpPerspective. In case you want to keep your previous behaviour set it to False. This warning will disappear in kornia > v0.6.
  warnings.warn(message)
Traceback (most recent call last):
  File "stitch_benchmark.py", line 171, in <module>
    data_warp_l = warp_image(img_l,M_l)
  File "stitch_benchmark.py", line 145, in warp_image
    data_warp: torch.tensor = kornia.warp_perspective(data, M, dsize=(h, w))
  File "/home/mobilrouter/.conda/envs/Mobil_Torch/lib/python3.8/site-packages/kornia/geometry/transform/imgwarp.py", line 115, in warp_perspective
    grid = transform_points(src_norm_trans_dst_norm[:, None, None], grid)
  File "/home/mobilrouter/.conda/envs/Mobil_Torch/lib/python3.8/site-packages/kornia/geometry/linalg.py", line 203, in transform_points
    raise TypeError(
TypeError: Tensor must be in the same device and dtype. Got trans_01 with (torch.float32, torch.float32) and points_1 with (torch.float32, torch.float32)

**Reply by shijianjian** [reply]
You need to put M as a CUDA tensor as well. 

@edgarriba The error message is buggy. It shall provide device dtype than dtype dtype.

**Reply by edgarriba** [reply]
we used to have a running time error message in `warp_perspective` about the devices and dtype but we  decided to remove and leave it to pytorch to complain. If it's necessary we could include it again. I'm seeing also that `transforms_points`  should be cleaned a bit too. @pentadotddot @shijianjian could you try a quick PR cleaning the assertions there and messaging ?

**Reply by edgarriba** [reply]
for proper reproducibility I would suggest the following code:
```python
device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
print(device)

# using opencv
img: np.ndarry = cv2.imread(...)
img = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)
img_t: torch.tensor = kornia.image_to_tensor(img, keepdim=False)  # BxCxHxW
img_t = img_t.float().to(device)

# or using torchvision
img_t: torch.Tensor = torchvisio.io.imread(...)[None] # BxCxHxW
img_t = img_t.float().to(device)

M: torch.Tensor = torch.eye(3, device=device)[None]

h,w = img_t.shape[2], img_t.shape[3]
# warp the original image by the found transform
imgt_warp: torch.tensor = kornia.warp_perspective(img_t, M, dsize=(h, w))

img_warp: np.ndarray = kornia.tensor_to_image(imgt_warp.byte())
```

**Reply by pentadotddot** [reply]
Thank you a lot! It is working now. It is indeed faster then the opencv, like 1.5X or so:)
